File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/CounterSignatureInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Integer
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.DLTaggedObject
import org.bouncycastle.asn1.cms.Attribute
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.cms.SignerId
import java.math.BigInteger

@Serializable
data class CounterSignatureInfo(
  val version: Int,
  val sid: SignerIdentifierInfo,
  val digestAlgorithm: AlgorithmInfo,
  val authenticatedAttributes: List<AttributeInfo>,
  val digestEncryptionAlgorithm: AlgorithmInfo,
  val encryptedDigest: TextualInfo,
  val counterSignature: TaggedObjectInfo?,
) : EncodableInfo {

  companion object {
    fun getInstance(sequence: DLSequence): CounterSignatureInfo {
      val iterator = sequence.iterator()

      val version = (iterator.next() as ASN1Integer).intValueExact()
      val signerSequence = (iterator.next() as DLSequence).toList()
      val signerIdentifierInfo = SignerIdentifierInfo(
        SignerId(
          X500Name.getInstance(
            signerSequence[0]
          ),
          BigInteger(
            (signerSequence[1] as ASN1Integer).toString()
          )
        )
      )
      val digestAlgorithm = AlgorithmInfo(
        AlgorithmIdentifier.getInstance(iterator.next())
      )

      val attributes = ((iterator.next() as DLTaggedObject).baseObject as DLSequence).map {
        AttributeInfo.getInstance(Attribute.getInstance(it))
      }

      val encryptionAlgorithm = AlgorithmInfo(
        AlgorithmIdentifier.getInstance(iterator.next())
      )

      val encryptedDigest = TextualInfo.getInstance(iterator.next())

      val counterSignature: TaggedObjectInfo? = if (iterator.hasNext()) {
        val obj = iterator.next()
        TaggedObjectInfo(
          (obj as DLTaggedObject).isExplicit,
          obj.tagNo,
          AttributeInfo.getInstance(Attribute.getInstance(obj.baseObject))
        )
      } else {
        null
      }

      return CounterSignatureInfo(
        version,
        signerIdentifierInfo,
        digestAlgorithm,
        attributes,
        encryptionAlgorithm,
        encryptedDigest,
        counterSignature
      )
    }
  }

  override fun toPrimitive(): ASN1Primitive =
    listOf(
      ASN1Integer(version.toLong()),
      sid.toPrimitive(),
      digestAlgorithm.toPrimitive(),
//      TaggedObjectInfo(false, 0, authenticatedAttributes.toPrimitiveList().toDLSet()).toPrimitive(),
      TaggedObjectInfo.getTaggedObject(
        false,
        0,
        authenticatedAttributes.toPrimitiveDLSet()
      ),
      digestEncryptionAlgorithm.toPrimitive(),
      encryptedDigest.toPrimitive(),
      counterSignature?.toPrimitive(),
    ).toDLSequence()
}

File: net/JetBrains.Serialization/src/Dataholders/CounterSignatureInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class CounterSignatureInfo : IEncodableInfo
{
  [JsonProperty("Version")] private int _version;
  [JsonProperty("Sid")] private SignerIdentifierInfo _sid;
  [JsonProperty("DigestAlgorithm")] private AlgorithmInfo _digestAlgorithm;

  [JsonProperty("AuthenticatedAttributes")]
  private List<AttributeInfo> _authenticatedAttributes;

  [JsonProperty("DigestEncryptionAlgorithm")]
  private AlgorithmInfo _digestEncryptionAlgorithm;

  [JsonProperty("EncryptedDigest")] private TextualInfo _encryptedDigest;
  [JsonProperty("CounterSignature")] private TaggedObjectInfo? _counterSignature;

  [JsonConstructor]
  public CounterSignatureInfo(
    int version,
    SignerIdentifierInfo sid,
    AlgorithmInfo digestAlgorithm,
    List<AttributeInfo> authenticatedAttributes,
    AlgorithmInfo digestEncryptionAlgorithm,
    TextualInfo encryptedDigest,
    TaggedObjectInfo? counterSignature = null)
  {
    _version = version;
    _sid = sid;
    _digestAlgorithm = digestAlgorithm;
    _authenticatedAttributes = authenticatedAttributes;
    _digestEncryptionAlgorithm = digestEncryptionAlgorithm;
    _encryptedDigest = encryptedDigest;
    _counterSignature = counterSignature;
  }

  public static CounterSignatureInfo GetInstance(DerSequence sequence)
  {
    var iterator = sequence.GetEnumerator();
    iterator.MoveNext();

    var version = ((DerInteger)iterator.Current).Value.IntValue;
    iterator.MoveNext();

    var signerSequence = ((DerSequence)iterator.Current).ToArray();

    var signerIdentifierInfo = new SignerIdentifierInfo(
      X509Name.GetInstance(signerSequence[0]),
      (DerInteger)signerSequence[1]
    );
    iterator.MoveNext();

    var digestAlgorithm = new AlgorithmInfo(AlgorithmIdentifier.GetInstance(iterator.Current));
    iterator.MoveNext();

    var attributes = ((DerSequence)((DerTaggedObject)iterator.Current).GetObject())
      .ToArray()
      .Select(it => AttributeInfo.GetInstance(Attribute.GetInstance(it)))
      .ToList();
    iterator.MoveNext();

    var encryptionAlgorithm = new AlgorithmInfo(AlgorithmIdentifier.GetInstance(iterator.Current));
    iterator.MoveNext();

    var encryptedDigest = TextualInfo.GetInstance((Asn1Encodable)iterator.Current);

    TaggedObjectInfo? counterSignature = null;
    if (iterator.MoveNext())
    {
      var obj = (DerTaggedObject)iterator.Current;
      counterSignature = new TaggedObjectInfo(
        obj.IsExplicit(),
        obj.TagNo,
        AttributeInfo.GetInstance(Attribute.GetInstance(obj.GetObject())));
    }

    return new CounterSignatureInfo(
      version,
      signerIdentifierInfo,
      digestAlgorithm,
      attributes,
      encryptionAlgorithm,
      encryptedDigest,
      counterSignature);
  }

  public Asn1Encodable ToPrimitive()
  {
    var primitiveValues = new List<Asn1Encodable>
    {
      new DerInteger(_version),
      _sid.ToPrimitive(),
      _digestAlgorithm.ToPrimitive(),
      TaggedObjectInfo.GetTaggedObject(false, 0, _authenticatedAttributes.ToPrimitiveDerSet()),
      _digestEncryptionAlgorithm.ToPrimitive(),
      _encryptedDigest.ToPrimitive()
    };

    if (_counterSignature != null)
    {
      primitiveValues.Add(_counterSignature.ToPrimitive());
    }

    return primitiveValues.ToDerSequence();
  }
}